mne/preprocessing/artifact_detection.py

```diff
@@ -333,3 +334,189 @@ def _annotations_from_mask(times, art_mask, art_name):
+        in ``raw``. If an events array, ignore any annotations in the raw data,
+        and operate based on these events only.
+    min_duration : float
+        The minimum time span in seconds between the offset of one and the
+        onset of the subsequent annotation (if ``events`` is ``None``) or
+        between two consecutive events (if ``events`` is an array) to consider
+        this period a "break". Defaults to 15 seconds.
+    t_start_after_previous, t_stop_before_next : float
+        Specifies how far the to-be-created "break" annotation extends towards
+        the two annotations or events spanning the break. This can be used to
+        ensure e.g. that the break annotation doesn't start and end immediately
+        with a stimulation event. If, for example, your data contains a break
+        of 30 seconds between two stimuli, and ``t_start_after_previous`` is
+        set to ``5`` and ``t_stop_before_next`` is set to ``3``, the break
+        annotation will start 5 seconds after the first stimulus, and end 3
+        seconds before the second stimulus, yielding an annotated break of
+        ``30 - 5 - 3 = 22`` seconds. Both default to 5 seconds.
+
+        .. note:: The beginning and the end of the recording will be annotated
+                  as breaks, too, if the period from recording start until the
+                  first annotation or event (or from last annotation or event
+                  until recording end) is at least ``min_duration`` seconds
+                  long.
+
+    ignore : iterable of str | None
+        Annotation descriptions starting with these strings will be ignored by
+        the break-finding algorithm. The string comparison is case-insensitive,
+        i.e., ``('bad',)`` and ``('BAD',)`` are equivalent. By default, all
+        annotation descriptions starting with "bad" and annotations
+        indicating "edges" (produced by data concatenation) will be
+        ignored. If ``None``, take all existing annotations into account.
+        If ``events`` is passed, this parameter has no effect.
+    %(verbose)s
+
+    Returns
+    -------
+    break_annotations : mne.Annotations
+        The break annotations, each with the description ``'BAD_break'``. If
+        no breaks could be found given the provided function parameters, an
+        empty `~mne.Annotations` object will be returned.
+
+    Notes
+    -----
+    .. versionadded:: 0.24
+    """
+    from mne.io import BaseRaw
+    _validate_type(item=raw, item_name='raw', types=BaseRaw, type_name='Raw')
+    _validate_type(item=events, item_name='events', types=(None, np.ndarray))
+
+    if min_duration - t_start_after_previous - t_stop_before_next <= 0:
+        raise ValueError(
+            f'The result of '
+            f'min_duration - t_start_after_previous - t_stop_before_next '
+            f'must be greater than 0, but it is: '
+            f'{min_duration - t_start_after_previous - t_stop_before_next}'
+        )
```

Comment:
Why can't I have a `min_duration` of 1 sec with `t_start_after_previous = t_start_before_next = 5`? Imagine I have events at 0, 10.5, and 30 sec, I would expect this function *will not* to annotate nothing between 0 and 10.5 (because the potential 0.5-sec annotation doesn't meet the `min_duration` threshold) but *will* add an annotation from 15.5 to 25 sec.

Reply:
It was in fact a conscious design choice I made, but I'm willing to discuss it.

So the parameter `min_duration` applies to the **break periods in the data**. This is essentially being used in a first pass: we're simply looking at the data (or actually, at its annotations or events, for that matter), checking if there are any periods of duration `min_duration` where no experimental events occurred. **These are break periods**.

Now, in a subsequent pass, we create the **break annotations**. Since one will typically not want them to extend all the way to the surrounding experimental events, we set `t_start_after_previous` and `t_stop_before_next` to control just how much of the break period will be "filled up" by the break annotation.

To me it felt more natural to do it this way:
- find breaks of sufficient duration
- fill up those breaks with annotations

It could be that `min_duration`, here, is a poor name choice or it needs better documentation.

I'm a little concerned now that it obviously got you confused, so I'd like to discuss the naming, docs, and functionality with again before we actually merge this …

@cbrnr @sappelhoff @agramfort WDYT?

Reply:
> So the parameter min_duration applies to the break periods in the data. 

This is consistent with my interpretation

> we're simply ... Now, in a subsequent pass ... To me it felt more natural to do it this way:

It sounds like you're talking about *how you've implemented* the code, let's first talk about how it should work *conceptually*. We can implement it however we need to in order to make it work the way we think it should conceptually.

To clarify, do you still think it is correct behavior when you have events at 0, 10.5, and 30 seconds, with `min_duration=1, t_start_after_previous=5, t_start_before_next=5`? Because to me it makes sense that this should create a single BAD_BREAK annotation from 15.5 to 25 seconds rather than raise an error. I don't see any reason to restrict `min_duration` to be less than the total length of the "buffer" that we put between the annotations and actual data segments.

Reply:
I think this is a reasonable behavior and I found the docstring to be clear and understandable. One improvement may be to write two sentences between the one-line initial docstr and the "parameters" documentation --> these two sentences could clarify that this is a two step procedure, step 1 uses min_duration, step 2 uses t_start... and t_stop...

Reply:
... in other words, to me it makes sense to think of the "breaks" as being any gap between events/annotation bounds that is at least `min_duration` long *and* `t_start_after_previous` after the previous *and* `t_stop_before_next` before the next event. This boolean logic is much more natural and simpler to parse as a user than needing to think about how we actually implemented a given algorithm, even if that algorithm is well described. And my example of 0/10.5/30 to me is a compelling use case for why the current implementation does not make sense...

Reply:
Ahh I think I see, I'm misunderstanding/wrong about what `min_duration` is -- it's not the minimum duration of the annotation that will be added but of the total break time. Let me look again at the docstring and see if what's here is okay and I just misread it!

Reply:
I could be wrong, but I think there's still a misunderstanding between @hoechenberger and @larsoner.  I think @hoechenberger is saying "`min_duration` refers to length of a break period, i.e. the *total time* between events (before accounting for `t_start_before_prev` and `t_stop_before_next`), whereas I think @larsoner is thinking of `min_duration` as the minimum duration *of what ends up getting annotated*.

To me the naming is problematic.  Any parameter called `min_duration` should IMO definitely refer to the minimum duration *of the resulting annotations* (this is, I think, how @larsoner interpreted it).  If it turns out to make more sense to parameterize the duration between events (rather than the duration of resultant annotations) then we definitely need a new name for that parameter.

Reply:
Maybe we could call it `min_break_duration` so it's clear it's referring to the break and not the annotation itself

Reply:
> Maybe we could call it `min_break_duration` so it's clear it's referring to the break and not the annotation itself

Sounds good to me!

I just spent 5 min trying to wrap my head around the other possibility – letting the parameter define the minimum duration of the **resulting** annotation. And it just doesn't work out in my mind … might just be me, though.

This just to say: for now, I think I prefer the current behavior, and would opt to rename the parameter to the suggested name.

Reply:
> I just spent 5 min trying to wrap my head around the other possibility – letting the parameter define the minimum duration of the **resulting** annotation. And it just doesn't work out in my mind … might just be me, though.

to me, it's more natural to think of this problem as:

1. how long of a buffer should I leave after an event before ignoring data?
2. how long before the next event do I need to start keeping data again?
3. if the remaining time in between is shorter than X, don't bother annotating it.

I think this indicates that we really need to get the param names and docstring super clear.  For the way I'm thinking about it I might say `left_buffer`, `right_buffer`, `min_duration` (where `buffer` could also be `margin` or `padding` or `gap`) and another possibility is `pre_event_buffer` and `post_event_buffer`).  Those names might also work for @hoechenberger's API (where you specify minimum *break* rather than minimum *annotation*) if everyone prefers that, assuming `min_duration` becomes `min_break_duration` or (better?) `min_inter_event_duration`.

Reply:
I have renamed `min_duration` to `min_break_duration` and expanded the docs a little in dd014b2

Reply:
> @sappelhoff OTOH seems to agree with the current approach – if I'm not mistaken?

Yes, I agree with the current approach. When I record and analyze my data, I know approximately how long the breaks in the paradigm were. For example every 20 minutes, there was a 1 minute break. And I know that apart from these breaks my paradigm is fast-paced. So I know I can set `min_break_duration = 30`, and then care about tstart... and tstop. Seems way simpler to me, but apparently that's controversial, so more opinions would be good.

Reply:
To add a little context / argue for my version:  what I'm calling `left_buffer` and `right_buffer` are *always* going to be experiment-dependent, depending on how long your "trial" is / when during the trial you stamp events (i.e., do you include an end-of-trial event or not).  In contrast, the `min_annotation_duration` (as @larsoner and I conceived of it) is not experiment-dependent at all, and might have a sensible default like 1 second or 5 seconds.  People should still be able to control it, but it's possible to set a default that wouldn't break for certain experiment designs.  In @hoechenberger's proposed API, `min_break_duration` is confounded with experimental design, so there's no way to set a safe default (I don't think?)

Reply:
@drammock 
I suppose we're also thinking about different use cases here… I mostly agree with what @sappelhoff said: I would typically know that, for a given study I ran, breaks last, say, 1+ mins or so, and then I would just set `min_break_duration` to that value to capture those breaks, whereas …

> . In contrast, the `min_annotation_duration` (as @larsoner and I conceived of it) is not experiment-dependent at all, and might have a sensible default like 1 second or 5 seconds.

… this seems to indicate you're really thinking about capturing rather short event-free segments too? This is not what I initially had in mind when creating this function. So maybe this is also a reason why we have a bit of confusion here?

What I like about the `min_break_duration` approach is that I don't have to do any math to capture a (sufficiently long) break. It will cause headaches for short break periods, as you suggested, yes. But that's actually not what I typically consider a break in the paradigms I've worked with in the past

Reply:
I'm thinking of past experiments I've done where I allowed participants to self-determine break length between blocks.  So some breaks might be 15 seconds, others might be 5 minutes.  But either one might involve big nasty EEG noise due to the participant stretching or scratching their nose.